query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,219 @@
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The <code>QueryService</code> class implements the logic for querying a specific DDF-based site
+ * for changes to be replicated or harvested.
+ */
+public class QueryService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryService.class);
+
+  private final QueryServiceTools queryServiceTools;
+
+  private final ScheduledExecutorService executor;
+
+  /** The current configuration for the site being queried by this query service. */
+  private Site site;
+
+  private SiteQueue siteQueue;
+
+  /** The adapter to communicate with the given site. */
+  private NodeAdapter adapter;
+
+  @VisibleForTesting ScheduledFuture future;
+
+  /**
+   * Instantiates a new query service.
+   *
+   * @param site the configuration for the site to be queried by this query service
+   */
+  public QueryService(Site site, QueryServiceTools queryServiceTools) {
+    this(site, queryServiceTools, Executors.newSingleThreadScheduledExecutor());
+  }
+
+  public QueryService(
+      Site site, QueryServiceTools queryServiceTools, ScheduledExecutorService executor) {
+    this.queryServiceTools = queryServiceTools;
+    this.siteQueue = queryServiceTools.getQueueFor(site);
+    this.executor = executor;
+    update(site);
+  }
+
+  /**
+   * Updates the configuration for the site currently being queried by this query service.
+   *
+   * @param site the new site configuration
+   */
+  public void update(Site site) {
+    synchronized (this) {
+      this.site = site;
+      try {
+        this.adapter = queryServiceTools.getAdapterFor(site);
+      } catch (AdapterException e) {
+        LOGGER.info(
+            "Failed to create an adapter for {} stopping the query service for now."
+                + site.getUrl());
+        stop();
+      }
+    }
+  }
+
+  /** Starts this query service. */
+  public void start() {
+    if (future == null || future.isDone()) {
+      future =
+          executor.scheduleAtFixedRate(
+              this::query, 0, queryServiceTools.getPeriod(), TimeUnit.SECONDS);
+    }
+  }
+
+  /** Stops this query service. */
+  public void stop() {
+    executor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void query() {
+    if (!adapter.isAvailable()) {
+      LOGGER.debug("System at " + site.getUrl() + " is currently unavailable");
+      return;
+    }
+
+    Set<Filter> filters = queryServiceTools.activeFiltersFor(site).collect(Collectors.toSet());
```

Comment:
✏️ We should probably sort the filters by priority such that we do the highest one first.

Reply:
✏️ Make it `final`